connectionmanager/connectionmanager.go

```diff
@@ -59,72 +59,67 @@ type ProtocolHandler struct {
 
 var handlersByTypeMutex = &sync.Mutex{}
 var handlersByType = map[ws.ProtoType]*ProtocolHandler{}
-var reconnectIntervalSeconds = 5
 var DefaultPingWait = time.Minute
 
 func GetWriteTimeout() time.Duration {
 	return writeWait
 }
 
-func SetReconnectIntervalSeconds(i int) {
-	reconnectIntervalSeconds = i
+var a *expBackoff = &expBackoff{
+	maxInterval:  60 * time.Minute,
+	attempts:     0,
+	exceededMax:  false,
+	maxBackoff:   120 * time.Minute,
+	smallestUnit: time.Minute,
 }
 
 func connect(
 	proto ws.ProtoType,
 	serverUrl, connectUrl, token string,
-	retries uint,
 	ctx context.Context,
 ) error {
 	parsedUrl, err := url.Parse(serverUrl)
 	if err != nil {
 		return err
 	}
-
 	scheme := getWebSocketScheme(parsedUrl.Scheme)
 	u := url.URL{Scheme: scheme, Host: parsedUrl.Host, Path: connectUrl}
 
 	cancelReconnectChan[proto] = make(chan bool)
 	var c *connection.Connection
-	var i uint = 0
 	defer func() {
 		setReconnecting(proto, false)
 	}()
+
 	setReconnecting(proto, true)
-	for IsReconnecting(proto) {
-		i++
-		c, err = connection.NewConnection(u, token, writeWait, maxMessageSize, DefaultPingWait)
-		if err != nil || c == nil {
-			if retries == 0 || i < retries {
-				if err == nil {
-					err = errors.New(
-						"unknown error: connection was nil but no error provided by" +
-							" connection.NewConnection",
-					)
-				}
-				log.Errorf("connection manager failed to connect to %s%s: %s; "+
-					"reconnecting in %ds (try %d/%d); len(token)=%d", serverUrl, connectUrl,
-					err.Error(), reconnectIntervalSeconds, i, retries, len(token))
-				select {
-				case cancelFlag := <-cancelReconnectChan[proto]:
-					log.Tracef("connectionmanager connect got cancelFlag=%+v", cancelFlag)
-					if cancelFlag {
-						return nil
-					}
-				case <-ctx.Done():
-					return nil
-				case <-time.After(time.Second * time.Duration(reconnectIntervalSeconds)):
-					break
-				}
-				continue
-			} else if i >= retries {
-				return ErrConnectionRetriesExhausted
+	c, err = connection.NewConnection(u, token, writeWait, maxMessageSize, DefaultPingWait)
```

Comment:
The back-off should protect `connection.NewConnection`. For example:
```go
	if err := a.Backoff(ctx); err != nil {
		return err
	}
	c, err = connection.NewConnection(u, token, writeWait, maxMessageSize, DefaultPingWait)
```
Where `a.WaitForBackoff` would look something like:
```go
func (a *expBackoff) Backoff(ctx context.Context, proto ws.ProtoType) error {
	a.attempts++
	if a.attempts <= 1 {
		return nil
	}
	intrval := a.GetExponentialBackoffTime()
	select {
	case <-time.After(a.GetExponentialBackoffTime()):
	case <-cancelReconnectChan[proto]:
		return context.Canceled
	case <-ctx.Done():
		return ctx.Err()
	}
	return nil
}
```